text/0000-actors.md

```diff
@@ -0,0 +1,890 @@
+
+In a publish operation, if *any* of the receiver's queues does not have enough space for the message then
+the `post` calls return an `Err`or and *no* message is posted or cloned.
+
+## Extensions
+
+This section covers extensions to the API. They are *not* required to fulfill the actor model but are "nice to have".
+
+### Field-level `#[init]` attribute
+
+This extension allows initialization of actors in the `#[actors]` struct itself.
+The argument of the field-level `#[init]` attribute is used as the initial state of an actor.
+
+``` rust
+#[actors]
+struct Actors {
+    #[init(Actor::new())]
+    actor: Actor,
+}
+```
+
+The argument can be omitted: this results in the actor being initialized using the `Default` trait.
+
+``` rust
+#[derive(Default)]
+struct Actor { /* .. */ }
+
+#[actors]
+struct Actors {
+    #[init]
+    actor: Actor,
+}
+```
+
+### Opting out of memory sharing
+
+"No memory sharing" is a core aspect of the actor model.
+To enforce that aspect in the rest of the RTIC API,
+this extension proposes adding a `forbid` key-value argument to the `#[rtic::app]` attribute.
+The value of the argument is a list of features to disable; one of them being `shared_resources`.
+
+Disabling the `shared_resources` feature makes it a compiler error to add any resource to the `#[shared]` struct.
+
+``` rust
+#[rtic::app(forbid = [shared_resources])]
+mod app {
+    #[shared]
+    struct Resources {
+        resource: Resource,
+        //~^ error: `shared_resources` have been disabled
+    }
+}
+```
+
+### Opting out of software tasks
+
+Actors share many similarities with software tasks but are a "higher-level", or more encapsulated, construct so
+mixing actors and software tasks in the same application can be seen as working at the wrong abstraction layer.
+
+This extension proposes a mechanism to disable software tasks.
+The surface syntax reuses the `forbid` key-value argument proposed in the previous subsection.
+
+Disabling the `software_tasks` feature makes it a compiler error to declare any software task.
+
+``` rust
+#[rtic::app(forbid = [software_tasks])]
+mod app {
+    #[task]
+    fn software(cx: software::Context) { /* .. */ } //~ error
+
+    #[task(binds = TIMER)]
+    fn hardware(cx: hardware::Context) { /* .. */ } //~ OK
+}
+```
+
+Hardware tasks are still allowed when `forbid = [software_tasks]` is used.
+Hardware tasks act as inputs to the actor network so most "real world" actor-based applications will use them alongside the `#[actors]` feature.
+
+### Memory watermarks
+
+This extension proposes adding an API to retrieve the highest observed queue "occupancy": a *memory watermark* API.
+The goal is to be able to measure the actual memory usage of the message queues internally used by actors.
+
+> Side note: although the memory watermark feature is presented in the context of this actor proposal, a similar API could be added to regular software tasks.
```

Comment:
This functionality seems generally desired: https://github.com/rtic-rs/cortex-m-rtic/issues/428
:+1: